### DISCUSSION TITLE
fetchart: Add an option to fetch covers from metadata sources

### DISCUSSION THREAD

**Discussion by evur** [beetbox/beets#4676] [Ideas]
There are instances where covers are only available from metadata sources like MusicBrainz and Discogs and are not present in the currently implemented sources.
This also allows for more precise fetching of covers by using release IDs instead of album titles.

## Fetching covers from metadata sources which are not supplying the metadata
Should a cover not be present on the metadata source of the release, other metadata sources can be used too, if their IDs are present.

Sometimes links to other metadata sources and release IDs can be retreived from the metadata source.
Such a feature is currently being worked on in PR #4708.
I recommend hinting at enabling this feature and mentioning that the IDs of the platforms to be fetched covers from must be present. This is an attempt to prevent users trying to troubleshoot why fetchart is only fetching covers from the current metadata source.

**Comment by arsaboo** [comment]
Beyond adding additional sources, it would be nice to extend the `embedart` plugin to accept an image file (link) to update the cover.

**Comment by JOJ0** [comment]
Hi @evur Discogs would be an awesome source for coverart because it has a cover for almost any release you find on there. I think I know were you are going with "Discogs ID can be retrieved from MB data" but please elaborate. Cheers!

**Comment by evur** [comment]
Hi @JOJ0, it's cool to see that you're taking interest in the idea!

> I think I know were you are going with "Discogs ID can be retrieved from MB data" but please elaborate.

I meant that if an album's data is being fetched from MB, the plugin can also try to fetch the cover from Discogs by getting the Discogs release id from the MB data.

**Comment by JOJ0** [comment]
Ok if I understand correctly you mean by fetching this information here, usually to find bottom right on an MB release page:

![image](https://user-images.githubusercontent.com/2733783/221170806-883b38bb-6014-4fc8-9cb4-8f4d0901a3ee.png)

Frotunately this information is available via the MusicBrainz API, only that currently beets doesn't fetch it. I started to work on that and actually have an already working solution to get the Discogs ID, as well as Spotify, Deezer and Bandcamp ID's, provided they are available with the MB relase, and the user enabled fetching it in beets config.yaml. They will be saved to the fields/flexible fields in the beets library.
This is the PR: https://github.com/beetbox/beets/pull/4474

I've recently set back the mentionend PR to draft since I figured that it might be easier to review for maintainers if I split it in two pieces. This is the first one that only handles to split out ID extraction things _only_ and should be merged first: https://github.com/beetbox/beets/pull/4633

If we handle do to that I'll be happy to make the former PR review ready again :-)

Hope that gives perspective on the timely order of things so far.

If we finish all that, would you be up for submitting the actual feature to `fetchart`` Since I'm not familiar with that plugin and can  predict that I won't find time to get into it. I'll find other equally important things in this project to do :-)

But yeah, again: Great feature idea! A lot of people would like that I think, including me!

**Reply by evur** [reply]
> Ok if I understand correctly you mean by fetching this information here, usually to find bottom right on an MB release page:

Yep, that's what I meant.

> If we finish all that, would you be up for submitting the actual feature to `fetchart`` Since I'm not familiar with that plugin and can predict that I won't find time to get into it. I'll find other equally important things in this project to do :-)

I already tried to implement it, but I quickly noticed that the Discogs code and authentication is tied to the Discogs plugin and must first be seperated. (see https://github.com/beetbox/beets/discussions/4680#discussioncomment-5074564)

But when the code is seperated, I would be happy to work on a PR.

**Comment by arsaboo** [comment]
Since we already have a few of these autotag sources (MB, Discogs, Spotify, etc.) that already provide cover art, why not use the images they provide as a starting point? The current approach of searching for cover art often leads to wrong images, especially if the album name is very common (e.g., Lucky, Bad Boy).  

Is it possible to call `embedart` from within the plugin and supply the image? I have an open PR https://github.com/beetbox/beets/pull/4719 that adds the ability to supply an image URL to the embedart plugin.
